File: day2.py
from aoc import readinput

sample = """forward 5
down 5
forward 8
up 3
down 8
forward 2
""".splitlines()

puzzle = sample
puzzle = readinput(2)

instructions = list()

for line in puzzle:
    direction, distance = line.split()
    distance = int(distance)
    instructions.append((direction,distance))

forward = 0
depth = 0

for direction, distance in instructions:
    if direction == 'forward':
        forward += distance
    elif direction == 'down':
        depth += distance
    elif direction == 'up':
        depth -= distance
    else:
        raise RuntimeError(f"unexpected {direction}")

print("#1", forward*depth)

forward = 0
depth = 0
aim = 0

for direction, distance in instructions:
    if direction == 'forward':
        forward += distance
        depth += distance*aim
    elif direction == 'down':
        aim += distance
    elif direction == 'up':
        aim -= distance
    else:
        raise RuntimeError(f"unexpected {direction}")

print("#2", forward*depth)

File: day2.js
"use strict";

const fs = require("fs");

function part1(steps) {
  let depth = 0;
  let pos = 0;
  for (const { direction, dist } of steps) {
    switch (direction) {
      case "forward":
        pos += dist;
        break;
      case "down":
        depth += dist;
        break;
      case "up":
        depth -= dist;
        break;
    }
  }
  return depth * pos;
}

function part2(steps) {
  let aim = 0;
  let depth = 0;
  let pos = 0;
  for (const { direction, dist } of steps) {
    switch (direction) {
      case "forward":
        pos += dist;
        depth += dist * aim;
        break;
      case "down":
        aim += dist;
        break;
      case "up":
        aim -= dist;
        break;
    }
  }
  return depth * pos;
}

const sample = `forward 5
down 5
forward 8
up 3
down 8
forward 2
`
  .split("\n")
  .filter((x) => Boolean(x))
  .map((line) => {
    const parts = line.split(" ");
    return { direction: parts[0], dist: Number(parts[1]) };
  });

console.log("sample1", part1(sample));
console.log("sample2", part2(sample));

fs.readFile("input2", "utf8", (err, data) => {
  if (err) {
    console.error(err);
    return;
  }
  const steps = data
    .split("\n")
    .filter((x) => Boolean(x))
    .map((line) => {
      const [direction, num] = line.split(" ");
      const obj = { direction, dist: Number(num) };
      return obj;
    });
  console.log("part1", part1(steps));
  console.log("part2", part2(steps));
});
